providers/efa/efa.c

```diff
@@ -57,6 +57,8 @@ static struct verbs_context *efa_alloc_context(struct ibv_device *vdev,
 
 	cmd.comp_mask |= EFA_ALLOC_UCONTEXT_CMD_COMP_TX_BATCH;
 	cmd.comp_mask |= EFA_ALLOC_UCONTEXT_CMD_COMP_MIN_SQ_WR;
+	strncpy((char *)cmd.userspace_ver, PACKAGE_VERSION,
```

Comment:
I'm not sure about it, but probably this is wrong version to report.
There are situations where provides/efa version will be different from rdma-core package_version.
For example, how well will it work for static compilied libraries?

Reply:
Is rdma-core version reall valuable anyhow? I tought most of the tricky code in EFA was part of libfabric?

Reply:
Don't you mean dynamically compiled libraries? How could the rdma-core package version be different than the EFA provider's if it's compiled statically?

What do you have in mind, passing
https://github.com/linux-rdma/rdma-core/blob/bdddadf6c8069a95ec355e1d7ce78fbee4fa71b0/providers/efa/CMakeLists.txt#L2
instead?

Reply:
> Don't you mean dynamically compiled libraries? How could the rdma-core package version be different than the EFA provider's if it's compiled statically?
> 
> What do you have in mind, passing
> 
> https://github.com/linux-rdma/rdma-core/blob/bdddadf6c8069a95ec355e1d7ce78fbee4fa71b0/providers/efa/CMakeLists.txt#L2
> 
> instead?

I mean that in theory you can compile rdma-core with -DENABLE_STATIC=1 and take the libefa.a file to your own hello_world application, but still use libibverbs.so.

Reply:
@jgunthorpe Well, it's both.
Having the libfabric version would clearly be nice too, but it's not necessarily instead of the rdma-core version (especially since not all apps necessarily run over libfabric).

Libfabric will require a weird direct verb to pass its version, that's a topic for another day 😄 .

Reply:
Well, I just worry about where this is going. Is it even appropriate to report this kind of information through rdma? Can't you have other agents to collect this?

Reply:
@rleon I see, isn't it possible to have two different .so versions of rdma-core and efa provider as well?

So something like this (might be missing a PARENT_SCOPE for EFA_VERSION):
```
diff --git a/buildlib/config.h.in b/buildlib/config.h.in
index c5b0bf55709d..9aa92d294bbc 100644
--- a/buildlib/config.h.in
+++ b/buildlib/config.h.in
@@ -9,6 +9,7 @@
 #define HAVE_BUILTIN_CLZL 1
 
 #define PACKAGE_VERSION "@PACKAGE_VERSION@"
+#define EFA_VERSION "@EFA_VERSION@"
 
 // FIXME: Remove this, The cmake version hard-requires new style CLOEXEC support
 #define STREAM_CLOEXEC "e"
diff --git a/providers/efa/CMakeLists.txt b/providers/efa/CMakeLists.txt
index ba0032a12c43..5815564318a7 100644
--- a/providers/efa/CMakeLists.txt
+++ b/providers/efa/CMakeLists.txt
@@ -1,5 +1,7 @@
+set(EFA_VERSION "1.1.${PACKAGE_VERSION}")
+
 rdma_shared_provider(efa libefa.map
-	1 1.1.${PACKAGE_VERSION}
+	1 ${EFA_VERSION}
 	efa.c
 	verbs.c
 )
```

Sounds good?

Reply:
> I mean that in theory you can compile rdma-core with -DENABLE_STATIC=1 and take the libefa.a file to your own hello_world application, but still use libibverbs.so.

No, you can't, it is blocked. All must be matched.

Reply:
I don't disagree, this has the potential to go too far in regards to what should be reported through rdma and what shouldn't.

But specifically talking about rdma-core/provider version, it is very tightly coupled with the EFA driver (as opposed to libfabric/any other app), it's two pieces of the same driver so the userspace version is actually needed to get the full picture.

Reply:
> > I mean that in theory you can compile rdma-core with -DENABLE_STATIC=1 and take the libefa.a file to your own hello_world application, but still use libibverbs.so.
> 
> No, you can't, it is blocked. All must be matched.

You are assuming that rdma-core is clean and received from distro. It is not always true.
It can be user-compiled and installed.

Reply:
No, our build system mangles all private symbols in the lib.a files to include the package version. It is deliberately made impossible to mix and match, and it is completely impossible for a static provider to link to a dynamic libibverbs at all.

$ nm lib/libmlx4.a
                 U rdmacore33_0_ibv_cmd_modify_cq

Reply:
Thanks

Reply:
Thanks Leon and Jason, I'm leaving the patch as is.
Let me know if anything else needs fix.